### ISSUE TITLE
Dynamic scheduled tasks

### ISSUE THREAD

**Issue by wollistik** [rails/solid_queue#186]
Hi @rosa,
I really appreciate the work you have done here and I am eager to switch over to solid_queue. There is only one thing left, which is holding me back.

Other gems like resque-scheduler or sidekiq-scheduler offer the ability to dynamically add or remove tasks to the schedule (see https://github.com/resque/resque-scheduler?tab=readme-ov-file#dynamic-schedules). Since everything seems to be stored in the database for solid_queue, this should be quite easy to achieve. 

Maybe it is only a documentation issue and it is already possible (this would be awesome 🤞 ).
Happy to hear from you!

**Comment by rosa** [comment]
@wollistik, no, this is not supported yet, you didn't miss it 😅 

This is something we can consider adding for sure. May I ask what's your use case for it?

**Comment by wollistik** [comment]
Hi @rosa,
we have some cron like jobs, where the user is able to

* Disable/Enable running the job
* Change the schedule on when and how often the job is performed

Therefore the dynamic scheduling feature was really nice to implement these requirements.

But after I thought about this the last two days, I might be completely redesign this feature and come up with a different solution, because it was rarely used and still required some quirks to get it working.

**Comment by rosa** [comment]
Ahh got it! That makes sense 👍  Yes, I thought a bit more about it yesterday and this is something I want to implement eventually 😊

**Comment by abrunner94** [comment]
Maybe this helps, but the way I implemented dynamic cron schedules is by having one job run every minute and check if the cron schedule matches the one stored in my database table, with my table containing cron expressions. I use Fugit to check if the cron matches the current time.

**Comment by AquisTech** [comment]
@abrunner94 Even I have done a similar thing. Mine is not exactly the cron like scheduling but I am having different schedules.
First task is scheduled to be executed after 15 days and then onwards every alternate days.
So in the execution of first job I enqueue next job and in that job I keep enqueueing next jobs every 2 days as per condition.

**Comment by KellyBennett** [comment]
I will just add my use case as well in case it helps motivate the feature!

TL;DR: I empowered our product managers to be able to control schedules without needing to recruit an engineer and get a new version of the app deployed.


More details;

I have several recurring tasks which I'm currently managing with Sidekiq-scheduler. These tasks are used to give the users of our product "rewards" for various tasks they've completed over a given period of time. I want the product designers to be able to adjust this period of time without needing to recruit an engineer, and get a new version of the app deployed, so I built a page in our admin portal. The product designers need only enter a time/frequency using basic English, and I use Fugit to translate that into a cron time. When the record is saved, an active record hook is called which updates the Sidekiq-scheduler's schedule live, without requiring a new deployment.

Certainly it's a "Nice to have" feature, but it's something I already have implemented, and something the designers are used to having by now, so I can't very well take it away from them. So I'm holding off on migrating to Solid Queue until dynamic schedules are supported. 


Thanks so much for the awesome feature though! I'm very excited to switch over to it in the (hopefully) near future! ❤️

**Comment by jordancrawfordnz** [comment]
In the meantime, if you can settle for being able to configure this at boot time, just a reminder that you can template your `recurring.yml` file as it's loaded using `ActiveSupport::ConfigurationFile` (in the `SolidQueue::Configuration#load_config_from_file` method - https://github.com/rails/solid_queue/blob/main/lib/solid_queue/configuration.rb#L187C11-L193).

Personally I'll be using this to pull schedule from environment variables, but there's nothing stopping you pulling values out of the database (but changes would only take effect when the workers restart!).

e.g.:

```
default: &default
  sync:
    class: SyncJob
    schedule: <%= ENV.fetch("SYNC_SCHEDULE", "every hour") %>

development:
  <<: *default

production:
  <<: *default
```

**Comment by javierav** [comment]
My two cents here. This feature can be useful for any application that allows its users to control when certain tasks should run.

Imagine an application that allows you to decide when you want to receive an email summary of your activity. You can implement this in two ways:

1. A job that runs every minute to check if any user's email needs to be sent.
2. Delegate this responsibility to Solid Queue to run the job when appropriate.

If you have few users, the second option may be more optimal so you don't have to run a job every minute. If you have many users, delegating this responsibility to Solid Queue may not be the best option.

**Comment by rosa** [comment]
Hey all! Thanks everyone for chiming in. This is definitely on my backlog, I just haven't found the time to work on it but I hope to have some time soon. In the meantime, if someone wants to open a PR, I'd be happy to review it! I'm going to unassign myself so it's clear this is up for grabs. 

Like @jason-rutherford observed, the plumbing for this is there, via the `static` attribute in `recurring_tasks`. Right now all tasks are created as `static`. New dynamic tasks need to have `static` set to `false` so that they're not deleted when the schedule is loaded from the configuration file. Then, we just need a way to load the task on demand. Something like `SolidQueue::Scheduler::RecurringSchedule.add_dynamic_task(...)`, which would create the task and schedule it. I think that'd be all 🤔 Perhaps some renaming around to make clear the dynamic vs. static distinction.

**Comment by AdamFreemer** [comment]
@rosa forgive my possible naivety here, new to working with SolidQueue, but perhaps you can help me out with a question. I have Solid Queue set up on a Rails 7.2 app. My intended use case is a dynamic recurring job scenario, where users create ETL processes that have an initial setup job, then a recurring sync job based on requirements for each ETL process (dataset) that's managed by user set attributes in the app. I'm doing something like this right now:

`
      SolidQueue::RecurringTask.create!(
        {
              class_name: 'ScheduledEtlJob',
              arguments: [id, table_or_view_name],
              schedule: schedule_str,
              description: "#{period.capitalize} ETL sync for #{table_or_view_name} at #{start_time.strftime('%H:%M')}"
        }
      )
`
which creates a recurring job, cued in the database and viewable on my recurring jobs tab on /jobs. The recurring job schedules correctly, but it never runs. I've noticed there is no scheduler process running, just the other 3:
`
$ ps aux | grep queue
gitlab-+  232301 34.3  4.1 1047920 152832 pts/5  Sl+  00:15   0:08 solid-queue-supervisor(1.1.4): supervising 232316, 232320, 232325
gitlab-+  232316  0.8  3.9 1047984 146972 pts/5  Sl+  00:15   0:00 solid-queue-dispatcher(1.1.4): dispatching every 1 seconds
gitlab-+  232320  3.0  3.9 981356 146124 pts/5   Sl+  00:15   0:00 solid-queue-worker(1.1.4): waiting for jobs in *
`
Forgive my not understanding the current implementation of SolidQueue::Scheduler::RecurringSchedule, but should this theoretically work, or no?

If I stub in a job in recurring.yml, I'll see the scheduler process spin up, but otherwise it's not there. I wasn't sure if this is core to the problem, or perhaps I'm completely misunderstanding the state of Solid Queue and it's ability to run dynamically created recurring jobs.

**Comment by cupatea** [comment]
> Hey all! Thanks everyone for chiming in. This is definitely on my backlog, I just haven't found the time to work on it but I hope to have some time soon. In the meantime, if someone wants to open a PR, I'd be happy to review it! I'm going to unassign myself so it's clear this is up for grabs.

Hey @rosa thanks for clarifying on this, I opened a [PR](https://github.com/rails/solid_queue/pull/553) addressing this feature, please have a look

Thanks!

**Comment by lrjbrual** [comment]
Hi @rosa , any update on this? Is it ready to be merged into the main branch? I saw regarding this dynamic scheduling #553. we are developing app, that needs this, maybe this will help our issue of dynamic scheduling, because good_job as far as we know it is not available. thank you in advance

**Comment by rosa** [comment]
Hey @lrjbrual, I'm a bit swamped with other work, so I'm not sure when I'll be able to get to this one, I'm sorry!

**Comment by sauloefo** [comment]
In my case, I was looking for having a different schedule for the development environment so my first go to solution was to use the same syntax as in `database.yml`:
```
update_usd_exchange_rate: &update_usd_exchange_rate
  class: UpdateUsdExchangeRateJob
  schedule: <%= ENV.fetch("MY_SCHEDULE_VARIABLE") %>
```
But this syntax is not allowed in `recurring.yml`.
Then, reading the docs, I learnerd about the `SOLID_QUEUE_RECURRING_SCHEDULE` variable that allows me to specify a different `recurring.yml` file.

So I added to my local `.env` file this entry:
```
SOLID_QUEUE_RECURRING_SCHEDULE=tmp/recurring-dev.yml
```

And now I can have a different schedule from that in production (defined in `config/recurring.yml`) for my development environment.

Hope this help somebody in need.

**Comment by AndrewTehan** [comment]
Hi @rosa !
Just wanted to check in and ask if there are any updates on this feature?
Dynamic scheduling would be very helpful for our app.
Thanks so much for your work!

**Comment by rosa** [comment]
Hey @AndrewTehan, no updates, but I'm hoping I can schedule a block of time to work on this (#553) one and #142 soon.
